main.go

```diff
@@ -407,38 +418,106 @@ func (b *Backend) updateCallStats(t shortTraceMsg) {
+		for range ticker.C {
+			m.populate()
+		}
+	}()
+}
+
+func (m *multisite) populate() {
+	m.rwLocker.RLock()
+	defer m.rwLocker.RUnlock()
+
+	dspOrder := []col{colGreen} // Header
+	for i := 0; i < len(m.sites); i++ {
+		for range m.sites[i].backends {
+			dspOrder = append(dspOrder, colGrey)
+		}
+	}
+	var printColors []*color.Color
+	for _, c := range dspOrder {
+		printColors = append(printColors, getPrintCol(c))
+	}
+
+	tbl := console.NewTable(printColors, []bool{
+		false, false, false, false, false, false,
+		false, false, false, false, false,
+	}, 0)
+
+	cellText := make([][]string, len(dspOrder))
+	for i := range dspOrder {
+		cellText[i] = make([]string, len(headers))
+	}
+	cellText[0] = headers
 	for i, site := range m.sites {
 		for j, b := range site.backends {
+			b.Stats.Lock()
 			minLatency := "0s"
 			maxLatency := "0s"
 			if b.Stats.MaxLatency > 0 {
 				minLatency = fmt.Sprintf("%2s", b.Stats.MinLatency.Round(time.Microsecond))
 				maxLatency = fmt.Sprintf("%2s", b.Stats.MaxLatency.Round(time.Microsecond))
 			}
-			cellText[i*len(site.backends)+j][0] = humanize.Ordinal(b.siteNumber)
-			cellText[i*len(site.backends)+j][1] = b.endpoint
-			cellText[i*len(site.backends)+j][2] = b.getServerStatus()
-			cellText[i*len(site.backends)+j][3] = strconv.FormatInt(b.Stats.TotCalls, 10)
-			cellText[i*len(site.backends)+j][4] = strconv.FormatInt(b.Stats.TotCallFailures, 10)
-			cellText[i*len(site.backends)+j][5] = humanize.IBytes(uint64(b.Stats.Rx))
-			cellText[i*len(site.backends)+j][6] = humanize.IBytes(uint64(b.Stats.Tx))
-			cellText[i*len(site.backends)+j][7] = b.Stats.CumDowntime.Round(time.Microsecond).String()
-			cellText[i*len(site.backends)+j][8] = b.Stats.LastDowntime.Round(time.Microsecond).String()
-			cellText[i*len(site.backends)+j][9] = minLatency
-			cellText[i*len(site.backends)+j][10] = maxLatency
+			cellText[i*len(site.backends)+j+1][0] = humanize.Ordinal(b.siteNumber)
+			cellText[i*len(site.backends)+j+1][1] = b.endpoint
+			cellText[i*len(site.backends)+j+1][2] = b.getServerStatus()
+			cellText[i*len(site.backends)+j+1][3] = strconv.FormatInt(b.Stats.TotCalls, 10)
+			cellText[i*len(site.backends)+j+1][4] = strconv.FormatInt(b.Stats.TotCallFailures, 10)
+			cellText[i*len(site.backends)+j+1][5] = humanize.IBytes(uint64(b.Stats.Rx))
+			cellText[i*len(site.backends)+j+1][6] = humanize.IBytes(uint64(b.Stats.Tx))
+			cellText[i*len(site.backends)+j+1][7] = b.Stats.CumDowntime.Round(time.Microsecond).String()
+			cellText[i*len(site.backends)+j+1][8] = b.Stats.LastDowntime.Round(time.Microsecond).String()
+			cellText[i*len(site.backends)+j+1][9] = minLatency
+			cellText[i*len(site.backends)+j+1][10] = maxLatency
+			b.Stats.Unlock()
 		}
 	}
+	console.RewindLines(len(cellText) + 2)
+	tbl.DisplayTable(cellText)
 }
 
 func (m *multisite) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 	w.Header().Set("Server", "SideKick") // indicate sidekick is serving
+	m.rwLocker.RLock()
+	defer m.rwLocker.RUnlock()
```

Comment:
Can we reduce the time we hold the lock? This seems like a huge bottleneck if somebody tries to grab a write lock.

AFAICT this will block everything until every single request has finished.

Reply:
Using `atomic.Pointer` is the cleanest and less error-prone method. See https://github.com/minio/sidekick/pull/113/files#r1673637298.